### ISSUE TITLE
Incorrect output of dot

### ISSUE THREAD

**Issue by sebasv** [rust-ndarray/ndarray#584]
I noticed very peculiar behavior of the dot product when the `ndarray-linalg` package is included in a project:

### Code to replicate:

```rust
extern crate ndarray;
#[cfg(test)]
mod tests {
    use ndarray::prelude::*;
    #[test]
    fn test_dgemv() {
        let a = arr2(&[[1.], [2.]]);
        let b = arr1(&[1., 2.]);
        println!("{:?}", a.t().dot(&b));
    }
}

```

### Output:
```
Intel MKL ERROR: Parameter 7 was incorrect on entry to cblas_dgemv.
[0.0] shape=[1], strides=[1], layout=C | F (0x3), const ndim=1
```

### dependencies:
```toml
[dependencies]
ndarray = "0.12.1"

[dependencies.ndarray-linalg]
version = "0.10"
features = ["intel-mkl"]
```

Note the lack of a panic. If I remove the `ndarray-linalg` dependency the output is as expected. 
IMO `ndarray` should not outsource the dot product to BLAS when another package requires BLAS, but that is secondary to the uncaught error and faulty returned value.

**Comment by sebasv** [comment]
I noticed the problem does not persist if I manually broadcast the shapes to match:
```rust
        println!("{:?}", &b.broadcast((1, 2)).unwrap().dot(&a));
```
```
[[5.0]] shape=[1, 1], strides=[1, 1], layout=C (0x1), const ndim=2
```

**Comment by bluss** [comment]
Thanks for the reproducer. While the method is called dot I guess in long form we need to call this operation matrix-vector multiplication (hence "gemv"), and it's avoided by making it a matrix-matrix multiplication in your workaround. (It dispatches based on dimensionality and not axis length).

**Comment by bluss** [comment]
I'm not sure we expect a panic here. Still need to fix the gemv call though.

First thing I thought about were the shapes, if the row vs column was an issue, not sure what the expected panic is?

`arr2(&[[1.], [2.]])` is a 2D array so it is an 2 × 1 matrix, transposed it's a 1 × 2 matrix.

`arr1(&[1., 2.])` is a 1D array so it can be a "row" or "column" depending on use. As the right hand operand of gemv, it is a row matrix, so in ndarray it has the shape 2 but a as a matrix in this operation it has the shape 2 × 1. Thus the matrix multiplication 1 × 2 by 2 × 1 is compatible with an 1 × 1 output.

The other thing, can we turn the blas error into a panic? I don't know an interface that lets us get an error code from blas? blas errors should not happen in ndarray those, since this is a bug in our glue code.

**Comment by bluss** [comment]
Parameter 6 is LDA (6 what I get when reproducing with openblas), and we fall afoul of this rule (M and N reversed since we go through cblas). This is due to assigning the 1 × 2 array strides [1, 1].

```
LDA is INTEGER
           On entry, LDA specifies the first dimension of A as declared
           in the calling (sub) program. LDA must be at least
           max( 1, m ).
```

**Comment by sebasv** [comment]
Yep, if I hack the function that returns the layout to alter on a transpose, the dot product works. Since the choice of C or F layout is a bit arbitrary if all but one dimensions have length 1 while row/column majority is not, I feel like this is the way to a solution. Posting a PR because that is probably easier for the discussion, but I am far from convinced that my solution is acceptable.

**Comment by bluss** [comment]
If the `layout=` print were totally accurate, it should say C | F for both before and after transpose, because the array is effectively one dimensional.

**Comment by sebasv** [comment]
Updated the PR per your suggestions, all tests pass locally, awaiting Travis results. I added a testcase as well. It might be good to include test cases for the n-dimensional case as well, but I don't know what behavior we should expect from BLAS in this case, the Intel and Netlib documentation did not clear this up for me.

**Comment by bluss** [comment]
Re error handling, what I can find leads to here, `xerbla` and error handling by replacing this function. I haven't looked into what that means and if we can do it in Rust, but I assume it means involving our glue code crates (blas-sys and similar). https://software.intel.com/en-us/mkl-developer-reference-c-xerbla

**Comment by bluss** [comment]
Fixed by sebasv in #585
